### PR TITLE
Fixed infinite loop on SyncDiskStage

### PR DESCRIPTION
**Pull request by cube0x8**
Previously, the synchronization process in `SyncFromDiskStage` would get the fuzzer stuck in an infinite loop if one of the synced files caused the target to crash or timeout. This issue arose because `state.last_time` was updated only after all files were evaluated. If a file caused the target to crash, `state.last_time` wasn't updated, leading the fuzzer to continuously sync the same buggy testcases, resulting in an infinite loop.

Now, files are added to `self.to_sync` for syncing, and `state.last_time` is updated only after each file has been successfully evaluated. Additionally, files are removed from `self.to_sync` before evaluation to prevent duplicate processing, ensuring that each file is processed exactly once.

### CODE REVIEW

File: libafl/src/stages/sync.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ use crate::{
     executors::{Executor, ExitKind, HasObservers},
     fuzzer::{Evaluator, EvaluatorObservers, ExecutionProcessor},
     inputs::{Input, InputConverter, UsesInput},
-    stages::{RetryRestartHelper, Stage},
+    stages::{HashSet, RetryRestartHelper, Stage},
```
Comment (domenukk) [inline libafl/src/stages/sync.rs:22]: Why `stages::HashSet`?

File: libafl/src/stages/sync.rs [hunk 2]
```diff
@@ -105,6 +106,19 @@ where
                     .unwrap()
                     .last_time = max_time;
             }
+            // Iterate over the paths of files left to sync.
+            // By keeping track of these files, we ensure that no file is missed during synchronization,
+            // even in the event of a target restart.
+            let to_sync = self.to_sync.clone();
```
Comment (domenukk) [inline libafl/src/stages/sync.rs:112]: Shouldn't these be stored in the metadata? If the input crashes, the set will be empty on reload, right?

File: libafl/src/stages/sync.rs (hunk 1 above)
Comment (domenukk) [reply libafl/src/stages/sync.rs:22]: In general, I'd just do a vec, it'll be a lot faster for a small number of entries

**Comment by cube0x8** [comment]
@domenukk you were right, I addressed your comments.
I tested the feature with one of my fuzzer, here the debug output:
```
[2024-05-17T08:49:25Z DEBUG libafl::stages::sync] Syncing from disk: "/tmp/sync"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/00180b99f6289d01c0aea16ceda498520878b3418201d8ba6a910d920fb63371"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/0002a94198b2807e84923726345236906f67a3587522bf038e16aae59403ac56"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/00074eabdadd5b323eb13a30d97d19eb6fc6a32ecf100a424d9286628e8ecf3a"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/97894bd4fb05809b"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/00171b0ded8800567428c680e179623f2a4030088be39b6c2c8c0e7b3ac6ff36"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/000c5e74650d306c6966ac12a53ff9467030945ac32cf0f1946e3ef5fbe66cb2"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/0005bdfdad5ab50058ad151d44302e7721a46120bf867f720b381a4d99a13b12"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/000499ba4440b44535e75aaa65a5b703ea649bf02917c98e009a7b270214b7c6"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/000108c4dd2db6cf22bfe6e306f999d2133e5804257fd18eb5719b4c7a6e7735"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/0009620b460efabbfd36a83f6dcaed1ed5d64163e6725689e42b80bfa01c0376"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/0005213d0dc6cb0b2e468c6ab4a45dd9bfbad2a426e945000c141c9b58bff696"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/00286ae2a70f34e9facec0003871bddb1ba938a1a7d9b45ba2eefbce2aa55dd0"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/000257b764f59c4e1a17a59a5ddd5c77a773337a358e570fc3301deee79c248e"
[2024-05-17T08:49:25Z INFO  libafl::stages::sync] Syncing file: "/tmp/sync/0004871c8fa3c02ffac4026c7b2c222b94063144ae034b9d9fea38b6cd535365"
[2024-05-17T08:49:25Z DEBUG libafl::stages::sync] Number of files to sync: 14
[2024-05-17T08:49:25Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/00180b99f6289d01c0aea16ceda498520878b3418201d8ba6a910d920fb63371"
[2024-05-17T08:49:25Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/0002a94198b2807e84923726345236906f67a3587522bf038e16aae59403ac56"
[2024-05-17T08:49:31Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/00074eabdadd5b323eb13a30d97d19eb6fc6a32ecf100a424d9286628e8ecf3a"
[2024-05-17T08:49:31Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/97894bd4fb05809b"
[2024-05-17T08:50:21Z ERROR libafl::executors::hooks::unix::unix_signal_handler] Timeout in fuzz run.
[2024-05-17T08:50:22Z DEBUG libafl::stages::sync] Syncing from disk: "/tmp/sync"
[2024-05-17T08:50:22Z DEBUG libafl::stages::sync] Number of files to sync: 10
[2024-05-17T08:50:22Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/00171b0ded8800567428c680e179623f2a4030088be39b6c2c8c0e7b3ac6ff36"
[2024-05-17T08:50:22Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/000c5e74650d306c6966ac12a53ff9467030945ac32cf0f1946e3ef5fbe66cb2"
[2024-05-17T08:50:28Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/0005bdfdad5ab50058ad151d44302e7721a46120bf867f720b381a4d99a13b12"
[2024-05-17T08:50:28Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/000499ba4440b44535e75aaa65a5b703ea649bf02917c98e009a7b270214b7c6"
[2024-05-17T08:50:34Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/000108c4dd2db6cf22bfe6e306f999d2133e5804257fd18eb5719b4c7a6e7735"
[2024-05-17T08:50:34Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/0009620b460efabbfd36a83f6dcaed1ed5d64163e6725689e42b80bfa01c0376"
[2024-05-17T08:50:34Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/0005213d0dc6cb0b2e468c6ab4a45dd9bfbad2a426e945000c141c9b58bff696"
[2024-05-17T08:50:40Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/00286ae2a70f34e9facec0003871bddb1ba938a1a7d9b45ba2eefbce2aa55dd0"
[2024-05-17T08:50:40Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/000257b764f59c4e1a17a59a5ddd5c77a773337a358e570fc3301deee79c248e"
[2024-05-17T08:50:46Z DEBUG libafl::stages::sync] Evaluating: "/tmp/sync/0004871c8fa3c02ffac4026c7b2c222b94063144ae034b9d9fea38b6cd535365"
[2024-05-17T08:50:58Z DEBUG libafl::stages::sync] Syncing from disk: "/tmp/sync"
[2024-05-17T08:50:58Z DEBUG libafl::stages::sync] Number of files to sync: 0
```

it looks like it's working fine

File: libafl/src/stages/sync.rs [hunk 3]
```diff
@@ -86,24 +92,53 @@ where
         state: &mut Z::State,
         manager: &mut EM,
     ) -> Result<(), Error> {
+        log::debug!("Syncing from disk: {:?}", self.sync_dir);
         let last = state
```
Comment (domenukk) [inline libafl/src/stages/sync.rs:96]: This could be cleaned up/sped up using `metadata_or_insert_with`, but I guess we can do it in another PR...

**Comment by tokatoka** [comment]
This got me into a infinite-loop inside load_from_directory...

**Comment by cube0x8** [comment]
Ok, I might have missed something. Do you have a scenario I can reproduce with?
